### PR TITLE
Layout 2020: Implement `clip: rect`

### PR DESCRIPTION
**Pull request by Manishearth**
This implements `clip: rect`

Unfortunately, none of the tests pass yet, they are all broken due to https://github.com/servo/servo/issues/27387

Additionally, currently `clip` does not seem to clip the element itself, only its children. I'm not quite sure what to do about that, I patterned this off of the code in the layout 2013 which handled clip immediately after scroll overflow.

### CODE REVIEW

**Comment by Manishearth** [comment]
The last commit fixes the problem, but I'm not sure if it's the right fix.

File: components/layout_2020/display_list/stacking_context.rs [hunk 1]
```diff
@@ -708,13 +710,35 @@ impl BoxFragment {
         builder.current_space_and_clip.spatial_id = builder.nearest_reference_frame;
     }
 
+    fn build_clip_frame_if_necessary(
+        &self,
+        builder: &mut StackingContextBuilder,
+        containing_block_info: &ContainingBlockInfo,
+    ) {
+        let clip = self.style.get_effects().clip;
+        if let ClipRectOrAuto::Rect(r) = clip {
```
Comment (SimonSapin) [inline components/layout_2020/display_list/stacking_context.rs:719]: https://drafts.csswg.org/css2/#clipping
> The clip property applies only to absolutely positioned elements. 

So this code should also check for `position: absolute` or `position: fixed`.

**Review by SimonSapin — COMMENTED** [review]
As discussed on Matrix https://drafts.csswg.org/css2/#clipping says

> Content that has been clipped does not cause overflow.

So this PR should also change `scrollable_overflow` computation for  to take the intersection of the clip rect with what would otherwise be the scrollable overflow.

**Comment by SimonSapin** [comment]
Let’s try this on top of https://github.com/servo/servo/pull/27399

@bors-servo try=wpt-2020

**Comment by Manishearth** [comment]
Sweet, a whole pile of PASS expected FAIL. Though i bet many of these would have been PASS expected FAIL before this PR, I plan to rebase and do a two-step test update.

**Comment by Manishearth** [comment]
PR updated with tests

**Comment by Manishearth** [comment]
So the clip-not-absolute-positioned tests are failing, because my impl seems to apply universally rather than to positioned elements

**Comment by Manishearth** [comment]
Fixed

**Comment by CYBAI** [comment]
More `PASS` tests 👀 

<details>

```
48 unexpected results that are NOT known-intermittents:
  â–¶ Unexpected subtest result in /css/css-transitions/properties-value-auto-001.html:
  â”” PASS [expected FAIL] clip auto(to) / events

  â–¶ Unexpected subtest result in /css/css-transitions/properties-value-auto-001.html:
  â”” PASS [expected FAIL] clip auto(from) / events


  â–¶ Unexpected subtest result in /css/cssom/serialize-values.html:
  â”” PASS [expected FAIL] clip: rect(1em, auto, 0.5px, 2000em)

  â–¶ Unexpected subtest result in /css/cssom/serialize-values.html:
  â”” PASS [expected FAIL] clip: auto

  â–¶ Unexpected subtest result in /css/cssom/serialize-values.html:
  â”” PASS [expected FAIL] clip: inherit


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-032.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-052.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-043.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-078.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-102.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-089.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-006.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-092.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-028.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-041.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-055.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-068.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-008.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-004.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-053.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-066.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-005.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-077.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-097.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-030.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-091.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-099.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-076.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-019.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-029.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-020.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-016.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-067.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-042.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-007.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-018.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-098.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-031.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-040.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-080.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-088.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-056.xht


  â–¶ PASS [expected FAIL] /_mozilla/css/clip_a.html


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-017.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-044.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-079.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-054.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-065.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-090.xht


  â–¶ PASS [expected FAIL] /css/CSS2/visufx/clip-064.xht


  â–¶ PASS [expected FAIL] /_mozilla/css/inline_absolute_hypothetical_clip_a.html

```

</details>

**Comment by Manishearth** [comment]
@bors-servo retry

```
  ▶ TIMEOUT [expected ERROR] /css/cssom/CSSStyleSheet-constructable-disallow-import.tentative.html
  │ 
  └ _RegisterApplication(), FAILED TO establish the default connection to the WindowServer, _CGSDefaultConnection() is NULL.
```

**Comment by Manishearth** [comment]
@bors-servo retry

```
  │ ALSA lib pcm.c:2564:(snd_pcm_open_noupdate) Unknown PCM default
  │ AL lib: (EE) ALCplaybackAlsa_open: Could not open playback device 'default': No such file or directory
  │ 0:00:00.928360543  7914 0x7f1420013d00 WARN                  openal gstopenalsink.c:634:gst_openal_sink_open:<autoaudiosink0-actual-sink-openal> error: Could not open device.
  │ 0:00:00.928375573  7914 0x7f1420013d00 WARN                  openal gstopenalsink.c:634:gst_openal_sink_open:<autoaudiosink0-actual-sink-openal> error: ALC error: Out of Memory
  │ 0:00:00.929041095  7914 0x7f1420013d00 WARN                     oss gstosssink.c:398:gst_oss_sink_open:<autoaudiosink0-actual-sink-oss> error: Could not open audio device for playback.
  │ 0:00:00.929052884  7914 0x7f1420013d00 WARN                     oss gstosssink.c:398:gst_oss_sink_open:<autoaudiosink0-actual-sink-oss> error: system error: No such file or directory
  │ 0:00:00.939149462  7914 0x7f1420006d40 FIXME                default gstutils.c:3980:gst_pad_create_stream_id_internal:<appsrc0:src> Creating random stream-id, consider implementing a deterministic way of creating a stream-id
  │ Cannot connect to server socket err = No such file or directory
  │ Cannot connect to server request channel
  │ jack server is not running or cannot be started
  │ JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
  │ JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
  │ 0:00:00.961377010  7914 0x7f1418002a70 WARN              jackclient gstjackaudioclient.c:381:gst_jack_audio_get_connection: could not create connection
  │ 0:00:00.961754417  7914 0x7f1418002a70 WARN                jacksink gstjackaudiosink.c:355:gst_jack_ring_buffer_open_device:<autoaudiosink1-actual-sink-jackaudio> error: Jack server not found
  │ 0:00:00.961826626  7914 0x7f1418002a70 WARN                jacksink gstjackaudiosink.c:355:gst_jack_ring_buffer_open_device:<autoaudiosink1-actual-sink-jackaudio> error: Cannot connect to the Jack server (status 17)
  │ 0:00:00.962657478  7914 0x7f1418002a70 WARN                 default oss4-property-probe.c:302:gst_oss4_property_probe_get_values:<autoaudiosink1-actual-sink-oss4> Can't open file descriptor to probe available devices: No such file or directory
  │ 0:00:00.962690030  7914 0x7f1418002a70 WARN                oss4sink oss4-sink.c:513:gst_oss4_sink_open:<autoaudiosink1-actual-sink-oss4> error: Could not open audio device for playback.
  │ 0:00:00.962697959  7914 0x7f1418002a70 WARN                oss4sink oss4-sink.c:513:gst_oss4_sink_open:<autoaudiosink1-actual-sink-oss4> error: system error: No such file or directory
  │ ALSA lib confmisc.c:767:(parse_card) cannot find card '0'
  │ ALSA lib conf.c:4568:(_snd_config_evaluate) function snd_func_card_driver returned error: No such file or directory
  │ ALSA lib confmisc.c:392:(snd_func_concat) error evaluating strings
  │ ALSA lib conf.c:4568:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
  │ ALSA lib confmisc.c:1246:(snd_func_refer) error evaluating name
  │ ALSA lib conf.c:4568:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
  │ ALSA lib conf.c:5047:(snd_config_expand) Evaluate error: No such file or directory
  │ ALSA lib pcm.c:2564:(snd_pcm_open_noupdate) Unknown PCM default
  │ AL lib: (EE) ALCplaybackAlsa_open: Could not open playback device 'default': No such file or directory
  │ 0:00:00.963222527  7914 0x7f1418002a70 WARN                  openal gstopenalsink.c:634:gst_openal_sink_open:<autoaudiosink1-actual-sink-openal> error: Could not open device.
  │ 0:00:00.963232102  7914 0x7f1418002a70 WARN                  openal gstopenalsink.c:634:gst_openal_sink_open:<autoaudiosink1-actual-sink-openal> error: ALC error: Out of Memory
  │ 0:00:00.963345964  7914 0x7f1418002a70 WARN                     oss gstosssink.c:398:gst_oss_sink_open:<autoaudiosink1-actual-sink-oss> error: Could not open audio device for playback.
  │ 0:00:00.963354746  7914 0x7f1418002a70 WARN                     oss gstosssink.c:398:gst_oss_sink_open:<autoaudiosink1-actual-sink-oss> error: system error: No such file or directory
  │ 0:00:00.963894965  7914 0x7f14180074f0 FIXME                default gstutils.c:3980:gst_pad_create_stream_id_internal:<appsrc1:src> Creating random stream-id, consider implementing a deterministic way of creating a stream-id
  │ 0:00:00.985385018  7914 0x7f1418007450 FIXME                default gstutils.c:3980:gst_pad_create_stream_id_internal:<appsrc2:src> Creating random stream-id, consider implementing a deterministic way of creating a stream-id
  │ 0:00:00.988049760  7914 0x7f1418007c00 FIXME                default gstutils.c:3980:gst_pad_create_stream_id_internal:<appsrc3:src> Creating random stream-id, consider implementing a deterministic way of creating a stream-id
  │ 0:00:00.989589679  7914 0x7f14180074a0 FIXME                default gstutils.c:3980:gst_pad_create_stream_id_internal:<appsrc4:src> Creating random stream-id, consider implementing a deterministic way of creating a stream-id
  │ 0:00:00.989827861  7914 0x7f1418007680 FIXME                default gstutils.c:3980:gst_pad_create_stream_id_internal:<appsrc5:src> Creating random stream-id, consider implementing a deterministic way of creating a stream-id
  │ 0:00:00.990594814  7914 0x7f141802fa80 FIXME                default gstutils.c:3980:gst_pad_create_stream_id_internal:<appsrc6:src> Creating random stream-id, consider implementing a deterministic way of creating a stream-id
  │ 0:00:00.992381306  7914 0x7f14180300a0 FIXME                default gstutils.c:3980:gst_pad_create_stream_id_internal:<appsrc7:src> Creating random stream-id, consider implementing a deterministic way of creating a stream-id
  └ 0:00:00.993424085  7914 0x7f1418007cf0 FIXME                default gstutils.c:3980:gst_pad_create_stream_id_internal:<appsrc8:src> Creating random stream-id, consider implementing a deterministic way of creating a stream-id
```

**Pull request merged**